url696d9a93c9_liaosiwei_bayesnetwork_tgz-pJ8-tutorial_TutorialJ8/src/com/alibaba/TrainV2.java

```diff
@@ -0,0 +1,338 @@
+
+/**
+ * User: sweeliao@gmail.com
+ * Date: 14-3-20
+ * Time: 下午4:26
+ */
+public class TrainV2 {
+    private static double [] clickProb = new double[2];
+    private static double [] carProb = new double[4];
+    private static double [] saveProb = new double[4];
+    private static double [] buyProb = new double[16];
+    private static JunctionTreeAlgorithm inferer = new JunctionTreeAlgorithm();
+    private static BayesNode a, b, c, d;
+    private static BayesNet net;
+
+    public static void createNetwork() {
+        net = new BayesNet();
+        a = net.createNode("a");
+        a.addOutcomes("true", "false");
+
+
+        b = net.createNode("b");
+        b.addOutcomes("true", "false");
+        b.setParents(Arrays.asList(a));
+
+
+        c = net.createNode("c");
+        c.addOutcomes("true", "false");
+        c.setParents(Arrays.asList(a));
+
+
+        d = net.createNode("d");
+        d.addOutcomes("true", "false");
+        d.setParents(Arrays.asList(a, b, c));
+
+        inferer.getFactory().setUseLogScale(true);
+        inferer.getFactory().setFloatingPointType(float.class);
+    }
+
+    public static void initNetwork() {
+        a.setProbabilities(clickProb);
+        b.setProbabilities(saveProb);
+        c.setProbabilities(carProb);
+        d.setProbabilities(buyProb);
+
+
+        inferer.setNetwork(net);
+    }
+
+    public static void setEvidence(Set<Integer> s) {
+        Map<BayesNode,String> evidence = new HashMap<BayesNode,String>();
+        for (Integer i: s) {
+            if (i == 0) evidence.put(a, "true");
+            else if (i == 2) evidence.put(b, "true");
+            else if (i == 3) evidence.put(c, "true");
+        }
+        inferer.setEvidence(evidence);
+    }
+
+    public static Boolean isExpendable() {
+        try {
+            double [] belifes = inferer.getBeliefs(d);
+/*            System.out.println(belifes[0] + " " + belifes[1]);*/
+            return belifes[0] > belifes[1];
+        }
+        catch (NumericalInstabilityException e) {
+            return false;
+        }
+    }
+
+    public static void main(String[] args) {
+        DataRead reader = new DataRead();
+        createNetwork();
+        if (!reader.read()) {
+            System.out.println("read file error");
+            return;
+        }
+        Map<String, Map<String, Map<Integer, HashSet<Integer>>>> table = reader.getData();
+
+        long totalSum = 0;
+
+        Set<String> allSet = new HashSet<String>();
+        Set<String> clickSet = new HashSet<String>();
+        Set<String> carSet = new HashSet<String>();
+        Set<String> saveSet = new HashSet<String>();
+        Set<String> buySet = new HashSet<String>();
+
+        for (String key1: table.keySet()) {
+            for (String key2: table.get(key1).keySet()) {
+                Map<Integer, HashSet<Integer>> one = table.get(key1).get(key2);
+                for (Integer i: one.keySet()) {
+                    HashSet<Integer> m = one.get(i);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.